File: Reverse_Linked_List.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode(int x) {
        val = x;
    }
}

class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null)
            return head;
        ListNode top = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return top;
    }
}

public class Reverse_Linked_List {
    public static void main(String[] args) {
        ListNode a = new ListNode(1);
        ListNode b = new ListNode(2);
        ListNode c = new ListNode(3);
        a.next = b;
        b.next = c;

        Solution sol = new Solution();
        ListNode ans = sol.reverseList(a);
        while (ans != null) {
            System.out.println(ans.val);
            ans = ans.next;
        }
    }
}

File: Reverse_Linked_List.go
package main

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

func reverseList(head *ListNode) *ListNode {
	var prev *ListNode = nil
	for head != nil {
		prev, head.Next, head = head, prev, head.Next
	}
	return prev
}

func main() {
	testCases := [][]int{{1, 2, 3, 4, 5}}
	for _, testCase := range testCases {
		PrintList(reverseList(BuildList(testCase)))
	}
}
